### PR TITLE
WIP: Normalized graph cut

### PR DESCRIPTION
**Pull request by tejus-gupta**
Implementation of recursive two-way normalized graph cut as described in this paper - ["Normalized Cuts and Image Segmentation"](https://people.eecs.berkeley.edu/~malik/papers/SM-ncut.pdf) by Shi and Malik. I had originally submitted a pull request at ImageSegmentation.jl but since this is more generally applicable, I was advised to submit a PR here. Since adjacency matrix for grid connectivity based graphs are sparse, some parts of the functions only work for sparse matrices. I will update the function to make it work for any array-like input.

### CODE REVIEW

**Comment by sbromberger** [comment]
Hi @tejus-gupta - thanks for this. If you're using `SimpleWeightedGraph`s, then you'll want to make this a PR to [SimpleWeightedGraphs.jl](https://github.com/JuliaGraphs/SimpleWeightedGraphs.jl). If it can be generalized to `AbstractGraph`s, then it belongs here.

File: src/graphcut/normalized_cut.jl [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+add_edge!(g, 2, 3, 1)
+add_edge!(g, 1, 3, 1)
+add_edge!(g, 4, 5, 1)
+add_edge!(g, 4, 6, 1)
+add_edge!(g, 5, 6, 1)
+add_edge!(g, 1, 6, 0.1)
+add_edge!(g, 3, 4, 0.2)
+
+labels = normalized_cut(g.weights, 0.1)
+```
```
Comment (sbromberger) [inline src/graphcut/normalized_cut.jl]: Our docstring convention is as follows:
```
"""
    command(arg1, arg2[, optarg1=val1]

Given an `arg1` and a value `arg2`, return a `bar`. If `optarg1` is specified, modify the return value by `baz`.

### Implementation Notes
something here

### Performance
Time complexity is ``O(n)```

### References
- ref1
- ref2
```

File: src/graphcut/normalized_cut.jl [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+"""
+function ncut{T<:Real}(W::SparseMatrixCSC{T,Int}, thres::Real = 0.01 ,num_cuts::Int = 10)
```
Comment (sbromberger) [inline src/graphcut/normalized_cut.jl]: If you're using abstract arguments (like Real), there's no specific need to typecast it. `thresh` can be untyped.

File: src/graphcut/normalized_cut.jl [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    #returns normalized cut score
+    function ncut_cost(mask)
```
Comment (sbromberger) [inline src/graphcut/normalized_cut.jl]: Embedded functions are generally not used in LightGraphs. Instead, pull this out of the top-level function and prefix it with a `_` to indicate it should not be used.

File: src/graphcut/normalized_cut.jl [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        cut_cost = 0
+        rows = rowvals(W)
+        vals = nonzeros(W)
+        for i = 1:n
+            for j in nzrange(W, i)
+                row = rows[j]
+                if mask[i] != mask[row]
+                    cut_cost += vals[j]/2
```
Comment (sbromberger) [inline src/graphcut/normalized_cut.jl]: You'll want to use a linter (like LanguageServer.jl's) to ensure your spacing is consistent.

File: src/graphcut/normalized_cut.jl [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        end
+        return cut_cost/sum(D*mask) + cut_cost/sum(D*(.~mask))
+    end
+
+    m, n = size(W)
+    D = Diagonal(vec(sum(W, 2)))
+
+    if m == 1
+        return [1]
+    end
```
Comment (sbromberger) [inline src/graphcut/normalized_cut.jl]: We shortcut non-mutating code when it's a one-liner:
```
m == 1 && return [1]
```

File: src/graphcut/normalized_cut.jl [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        end
+        W1 = sparse(I1, J1, V1)
+        W2 = sparse(I2, J2, V2)
+        labels1 = ncut(W1, thres, num_cuts)
+        labels2 = ncut(W2, thres, num_cuts)
+
+        labels = Vector{Int}(m)
+        n1 = maximum(labels1)
+
+        for i in 1:(m - sum(mask))
```
Comment (sbromberger) [inline src/graphcut/normalized_cut.jl]: Is it more efficient to calculate `sum(mask)` once and reuse that variable?

File: test/graphcut/normalized_cut.jl [hunk 7]
```diff
@@ -0,0 +1,24 @@
+@testset "Normalized Cut" begin
+
+    g = SimpleWeightedGraph(4)
+    add_edge!(g, 1, 2, 1)
+    add_edge!(g, 3, 4, 1)
+
+    labels = ncut(g.weights, 0.1)
```
Comment (sbromberger) [inline test/graphcut/normalized_cut.jl]: See the other test scripts: basically we loop through graphs of different eltypes:
```
gx = SimpleWeightedGraph(4)
add_edge!(gx, 1, 2, 1)
add_edge!(gx, 3, 4, 1)

for g in testgraphs(gx)
   labels = ncut(g.weights, 0.1)
   @test ....
end
```

File: src/graphcut/normalized_cut.jl [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        return [1]
+    end
+
+    #get eigenvector corresponding to second smallest eigenvalue
+    v = eigvecs(full(D-W), full(D))[:, 2]
+
+    #perform n-cuts with different partitions of v and find best one
+    min_cost = Inf
+    best_thres = -1
+    for t in linspace(minimum(v), maximum(v), num_cuts)
```
[the remark this answers is not in the document]
Comment (jpfairbanks) [reply src/graphcut/normalized_cut.jl]: As in `w = sort(v); (w[i+1] + w[i])/2`? That way you are guaranteed that all n cuts are distinct.

File: src/graphcut/normalized_cut.jl [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+    m, n = size(W)
+    D = Diagonal(vec(sum(W, 2)))
+
+    if m == 1
+        return [1]
+    end
+
+    #get eigenvector corresponding to second smallest eigenvalue
+    v = eigvecs(full(D-W), full(D))[:, 2]
```
[the remark this answers is not in the document]
Comment (tejus-gupta) [reply src/graphcut/normalized_cut.jl]: For my second test case, this is giving an error - `Base.LinAlg.ARPACKException("unspecified ARPACK error: -9999")`. 
You can see the detailed error message by ```Pkg.test("LightGraphs")```.

File: src/graphcut/normalized_cut.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (tejus-gupta) [reply src/graphcut/normalized_cut.jl]: From the paper - ```Currently, the search is done by checking l
evenly spaced possible splitting points, and computing the best Ncut among them.```. It wasn't clear to me exactly how they searched for the best partition.  https://github.com/scikit-image/scikit-image/blob/master/skimage/future/graph/graph_cut.py#L76 (line 210) Here they search evenly between min and max.

**Comment by sbromberger** [comment]
@jpfairbanks https://travis-ci.org/JuliaGraphs/LightGraphs.jl/builds/271017544#L541-L543

**Comment by tejus-gupta** [comment]
(continuing a discussion in one of the comments above, that thread was collapsed since the code is outdated)
```eigs``` is giving an error in my second test case.
```
D_W = [1.0 -1.0 0.0 0.0; -1.0 1.0 0.0 0.0; 0.0 0.0 1.0 -1.0; 0.0 0.0 -1.0 1.0]
D = [1.0 0.0 0.0 0.0; 0.0 1.0 0.0 0.0; 0.0 0.0 1.0 0.0; 0.0 0.0 0.0 1.0]
eigs(D_W, D, nev=2, which=:SR)
```
When ```D-W``` and ```D``` have the above values, I am getting this error:
 ```ERROR: Base.LinAlg.ARPACKException("unspecified ARPACK error: -9999")```.

**Comment by jpfairbanks** [comment]
I get the same error, I think it has to do with the fact that when given two matrix arguments ARPACK tries to solve the generalized eigenvalue problem, $D_Wx = lambda Dx$. This can be avoided by normalizing the matrix beforehand and using a standard eigenvalue problem. $D^-1 D_W x = lambda x$

**Comment by timholy** [comment]
If your inputs are symmetric, wouldn't it be better to do `L = cholfact(D, :L)` and solve
```
L^-1 D_W L^-1^T y = λy
```
for `y = L^T x`?

**Comment by jpfairbanks** [comment]
D is the diagonal matrix of weights so you don't need the cholfact, right?

**Comment by timholy** [comment]
I hadn't looked. Even easier then, just `sqrt.(D)`. But still best to keep things symmetric, it will use a more robust algorithm.

**Comment by sbromberger** [comment]
Is this PR still active? What do we want/need to do with it?

**Comment by tejus-gupta** [comment]
I changed to code use `eigs` with standard eigenvalue problem instead of generalized eigenvalue problem. `eigs(inv(D)*(D-W), nev=2, which=:SR)`

But I was still facing some issues -
```
D = [1.0 0.0; 0.0 1.0]
D_W = [1.0 -1.0; -2.0 2.0]
eigs(inv(D)*D_W)
```
Error: ArgumentError: input matrix A is too small. Use eigfact instead.

When eigenvalues are repeated, I sometimes don't get all the eigenvalues.
e.g,
```
D_W = [1.0 -1.0 0.0 0.0; -1.0 1.0 0.0 0.0; 0.0 0.0 1.0 -1.0; 0.0 0.0 -1.0 1.0]
D = [1.0 0.0 0.0 0.0; 0.0 1.0 0.0 0.0; 0.0 0.0 1.0 0.0; 0.0 0.0 0.0 1.0]
eigs(inv(D)*D_W)
```
output = ([2.0, 2.0, -4.39379e-17], [-0.562149 0.428939 0.694331; 0.562149 -0.428939 0.694331; 0.428939 0.562149 0.133806; -0.428939 -0.562149 0.133806], 3, 1, 4, [0.0, 0.0, 0.0, 0.0])

Sorry for the late update. I couldn't understand the above discussion since I don't know how Cholesky factorization works.

**Comment by timholy** [comment]
I submitted a PR to your fork, @tejus-gupta, to work around the ARPACK bug. One thing you'll need to check is the warning: it sometimes reduces the number of eigenvectors to 1, which I interpret as meaning the problem is degenerate. In that case you're not getting the 2nd-smallest eigenvalue, you're getting the only eigenvalue. I didn't go back and check the algorithm to determine the implications of this situation.

Sorry I didn't explain the above as carefully as I should have:
```
D_w x = λ D x
D = C' C   # This is the definition of the Cholesky factorization, assumes D is positive-definite
Hence
D_w x = λ C' C x
define
y = C x  (so x = inv(C) y)
hence
D_w inv(C) y = λ C' y
hence
A y = λ y    # an ordinary eigenvalue problem
where
A = inv(C)' D_w inv(C)
Then solve for x with x = inv(C) y
```
In this case, though, since `D` is diagonal the Cholesky factorization is trivial (just `C = sqrt.(D)`).

**Comment by sbromberger** [comment]
Hi @tejus-gupta @jpfairbanks @timholy - since this is well outside my area of expertise, I am relying on each of you to tell me when this is ready to merge. Thanks :)

**Comment by tejus-gupta** [comment]
@timholy thanks for explanation :smile:.

The warning saying that the number of eigenvalues is 1 seems to be a problem with `eigs`.
```
eigs( [1.0 -1.0; -1.0 1.0], nev=2, which=:SR)[2]
```
This problem has eigenvalues 2 and 0. The corresponding eigenvectors are λ * [-1, 1] and λ * [1, 1]. (I verified with numpy)

**Comment by jpfairbanks** [comment]
@timholy's comment and the current code reflect my understanding of the symmetric normalized laplacian eigenproblem `D^-1/2 (D-W) D^-1/2 y = lambda y` with `x = D^-1/2 y`.

I quibble with the evenly spaced thresholds for finding the optimal cut given the eigenvector. Because you can check all `n` possible cuts with some fancy counting. But I think we should get this merged and then offer that as an alternative later. For the purpose of consistency with the literature, this method of evenly spaced splits is worth including.

**Comment by jpfairbanks** [comment]
regarding the patch coverage, running the algorithm on a fully connected graph should give you the last 1 line of coverage.

**Comment by tejus-gupta** [comment]
@timholy What should I do with the problem that sometimes `eigs` is giving one eigenvalue when two eigenvalues should be returned?

**Comment by timholy** [comment]
https://github.com/JuliaLang/julia/blob/29eb4c67b424fdd393a326e186442c6fdd8743e6/base/linalg/arnoldi.jl#L179-L186 seems to suggest that for 2x2 one needs to use `eigfact`. I don't pretend to have dug deep enough to understand why, other than to say if I comment those lines out I get
```julia
julia> eigs( [1.0 -1.0; -1.0 1.0], nev=2, which=:SR)
ERROR: ARPACKException: unspecified ARPACK error: -3
Stacktrace:
 [1] aupd_wrapper(::Type, ::getfield(Base.LinAlg, Symbol("#matvecA!#218")){Array{Float64,2}}, ::getfield(Base.LinAlg, Symbol("##212#219")), ::getfield(Base.LinAlg, Symbol("##213#220")), ::Int64, ::Bool, ::Bool, ::String, ::Int64, ::Int64, ::String, ::Float64, ::Int64, ::Int64, ::Array{Float64,1}) at ./linalg/arpack.jl:51
 [2] #_eigs#211(::Int64, ::Int64, ::Symbol, ::Float64, ::Int64, ::Void, ::Array{Float64,1}, ::Bool, ::typeof(Base.LinAlg._eigs), ::Array{Float64,2}, ::UniformScaling{Bool}) at /home/tim/src/julia-1.0/base/linalg/arnoldi.jl:283
 [3] (::getfield(Base.LinAlg, Symbol("#kw##_eigs")))(::Array{Any,1}, ::typeof(Base.LinAlg._eigs), ::Array{Float64,2}, ::UniformScaling{Bool}) at ./<missing>:0
 [4] #eigs#181 at ./linalg/arnoldi.jl:90 [inlined]
 [5] (::getfield(Base.LinAlg, Symbol("#kw##eigs")))(::Array{Any,1}, ::typeof(eigs), ::Array{Float64,2}, ::UniformScaling{Bool}) at ./<missing>:0
 [6] #eigs#180 at ./linalg/arnoldi.jl:89 [inlined]
 [7] (::getfield(Base.LinAlg, Symbol("#kw##eigs")))(::Array{Any,1}, ::typeof(eigs), ::Array{Float64,2}) at ./<missing>:0
 [8] top-level scope
```
So clearly this is a case where the Julia devs are being careful.

So perhaps you can have a branch for the 2x2 case and then use `eigfact`?

**Comment by ViralBShah** [comment]
ARPACK does not handle the really tiny cases (smaller than roughly 10x10) because of the way its workspace is allocated. Perhaps we should just throw an error there or just call `eigfact` directly.

**Comment by timholy** [comment]
I think from the standpoint of the user, `eigs` means "give me the answer" and not "I was really hoping you'd use ARPACK for this computation." Calling `eigfact` and doing any necessary repackaging to return the result in the same form seems like the way to go.

**Comment by ViralBShah** [comment]
I agree, it should be from the user standpoint. Currently:
```
julia> eigs( [1.0 -1.0; -1.0 1.0], nev=2, which=:SR)
WARNING: Adjusting nev from 2 to 1
([-1.66533e-16], [-0.707107; -0.707107], 1, 1, 2, [0.0, 0.0])
```

**Comment by timholy** [comment]
I don't know what's up with the tests, but there is a merge conflict. Can you rebase?

**Comment by sbromberger** [comment]
I just fixed the merge conflict.

**Comment by timholy** [comment]
Strange, for me GitHub is still displaying this as a merge conflict.

**Comment by sbromberger** [comment]
Me too. Not sure why.

**Comment by sbromberger** [comment]
There we go.

**Comment by sbromberger** [comment]
Coverage missing for the conditional at https://codecov.io/gh/JuliaGraphs/LightGraphs.jl/src/6d9c9fd8f384b0d17ec69e28e3a525b79695593b/src/graphcut/normalized_cut.jl#L143...144

**Comment by jpfairbanks** [comment]
So that line is handling a corner case where the cut is either none of the vertices or all of the vertices, in which case you want to say everyone is in one group.

But I have a bigger problem:
```julia
julia> g = PathGraph(100)
{100, 99} undirected simple Int64 graph

julia> unique(normalized_cut(g, adjacency_matrix(g)))
1-element Array{Int64,1}:
 1

julia> unique(normalized_cut(g, adjacency_matrix(g), thres=1))
ERROR: BoundsError: attempt to access 2-element Array{Int64,1} at index [3]
Stacktrace:
 [1] getindex(::Array{Int64,1}, ::Int64) at ./array.jl:520
 [2] _recursive_normalized_cut(::SparseMatrixCSC{Float64,Int64}, ::Int64, ::Int64) at /Users/jfairbanks6/.julia/v0.6/LightGraphs/src/graphcut/normalized_cut.jl:161
 [3] _recursive_normalized_cut(::SparseMatrixCSC{Float64,Int64}, ::Int64, ::Int64) at /Users/jfairbanks6/.julia/v0.6/LightGraphs/src/graphcut/normalized_cut.jl:151 (repeats 4 times)
 [4] (::LightGraphs.#kw##normalized_cut)(::Array{Any,1}, ::LightGraphs.#normalized_cut, ::LightGraphs.SimpleGraphs.SimpleGraph{Int64}, ::SparseMatrixCSC{Int64,Int64}) at ./<missing>:0
```

I noticed that if you don't pass a `W` matrix explicitly, then you get a partition of everything in the same group, because the algorithm uses `weights(g)` as the default weight matrix. And `weights(g)` is `ones(nv(g), nv(g))` so you need to only access `W` on edges that exist. Or set a default value for weights as `adjacency_matrix(g)`, which is the matrix you want.

Even when you pass the correct `W` to the algorithm, the path graph should be partitioned into two groups `1:nv(g)/2` and `nv(g)/2:nv(g)`. This should be added as a test case for this algorithm.

**Comment by tejus-gupta** [comment]
The `sum(cut) == 0 || sum(cut) == m` condition never holds true because normalized cut cost is NaN in this case. The outer if-else makes sure that `ones(Int, m)` is returned.

**Comment by tejus-gupta** [comment]
The sum(cut) == 0 || sum(cut) == m condition never holds true because normalized cut cost is NaN in this case. The outer if-else makes sure that ones(Int, m) is returned.

**Comment by tejus-gupta** [comment]
The `sum(cut) == 0 || sum(cut) == m` condition never holds true because normalized cut cost is NaN in this case. The outer if-else makes sure that `ones(Int, m)` is returned.

**Comment by jpfairbanks** [comment]
@tejus-gupta, is there a calling mistake in the `PathGraph` example above? I would hope that `normalized_cut(PathGraph(n))` would just work since that graph has such an obvious partition.

**Comment by tejus-gupta** [comment]
There was a small mistake in the code. I have corrected it and added your example as a test case.

**Comment by jpfairbanks** [comment]
It looks fixed now, but here is another test case. Sorry for doing this painfully, but this algorithm is important.

```julia
g = blkdiag(CompleteGraph(25), CompleteGraph(10))
add_edge!(g, 1, nv(g)-1)
M = full(adjacency_matrix(g))
```

You can see from `eig(M)[2]` that spectral clustering should be able to separate this graph into the two obvious clusters. `1:25` and `26:nv(g)`

**Comment by tejus-gupta** [comment]
Its working for me. `normalized_cut(g, thres=0.02)` separated the graph into the two obvious clusters. With the default thres=0.01, the normalized cut for this split wasn't above the threshold and so graph wasn't partitioned.

Should `thres` be a required argument instead of keyword argument? If `thres` is a keyword argument, it suggests that the function will work fine without the argument and the argument allows the user to customize the function call. `thres=0.01` worked well enough to be default for images but it may not work well for all graphs.

**Comment by jpfairbanks** [comment]
This gets at a broader issue about algorithms that require choice of parameters. I would like to only offer a default value if we think that it will work for many input graphs. If there is basically no way to determine a good general purpose threshold, then we need to provide some guidance and say, you have to figure out a threshold  on your own.

Do you have some test images that you would like to use? If we could find a default value for `thres`  that works for the blkdiag graph, PathGraph, CompleteGraph, and some images. I would feel good about that threshold.

1. PathGraph should partition to 1:floor(n/2), ceil(n/2+1):n
2. CompleteGraph should not partition
3. Barbel graph blkdiag(CompleteGraph(n), CompleteGraph(m)) + 1 add_edge! should be 1:n,n:m+n
4. Images should give good patches.

These should be the test cases and then we have freedom to refine this algorithm as long as it gives the same results on these test cases.

**Comment by sbromberger** [comment]
What is the plan for this PR?

**Comment by tejus-gupta** [comment]
@jpfairbanks I don't we can have a threshold that would partition any PathGraph to 1:floor(n/2), ceil(n/2+1):n. For example, if I choose a threshold such that PathGraph with 2 vertices is split into two graphs with single vertex, then the algorithm can't stop after partitioning a PathGraph with 4 vertices into two graphs with 2 vertices each. It would also split the each of the graphs into graphs with single vertex.

The stopping criterion of the algorithm isn't the number of subgraphs, instead it would stop at a fixed subgraph size. Therefore I can set a threshold that would partition any PathGraph into subgraphs of fixed size.

I don't think there is any 'good' default value of the threshold. We should keep it as a required argument.

**Comment by jpfairbanks** [comment]
I'm not too worried about the n=2 case. 

I know that spectral partitioning for general graphs is hard, and having a good default is important. I don't want to let the perfect be the enemy of the good so we could let the threshold be a required argument and have the documentation say that it is critical to choose this based on your application and suggest a binary search over thresholds to choose a good one based on your application.

@sbromberger what do you think about that?

**Comment by sbromberger** [comment]
@jpfairbanks that sounds like a good plan. My preference would be to document the heck out of this, and then merge it. :)

S.

**Comment by jpfairbanks** [comment]
Cool. This means that the W argument has to go after the threshold argument. I am updating the tests now.

**Comment by jpfairbanks** [comment]
I just pushed changes to a branch `ncut` which can get merged into this PR in order to preserve this PR, or we can merge that and close this PR.

**Comment by sbromberger** [comment]
Let's preserve this one if possible.

Also, @jpfairbanks - could you add your review? I'll take a look from a style perspective but it needs some technical eyes on it.

**Comment by tejus-gupta** [comment]
@jpfairbanks You probably wanted to use `normalized_cut(g, t)` instead of `normalized_cut(g, 0.1)` on line 87 of test file. We would have to modify the test on line 89 also since the number of unique labels wouldn't be same for different thresholds.

**Comment by jpfairbanks** [comment]
@tejus-gupta that is exactly right, I wondered why I kept getting the same answer but didn't notice that error.

Once that is fixed we should merge it.

**Comment by jpfairbanks** [comment]
@tejus-gupta that is exactly right, I wondered why I kept getting the same answer but didn't notice that error.

Once that is fixed we should merge it.

**Comment by tejus-gupta** [comment]
I have fixed the test.

**Comment by sbromberger** [comment]
Many thanks to all of you for your persistence.

**Comment by sbromberger** [comment]
one last set of tests :)

**Comment by timholy** [comment]
Hooray! Thanks for persisting, @tejus-gupta!

**Pull request merged**